### ISSUE TITLE
경진대회 성능 최적화: 20분 내 RMSE 최소화 전략

### ISSUE THREAD

**Issue by iejob** [AIBootcamp14/upstageailab-ml-competition-ml-5#81]
**목표:** 현재 모델링 파이프라인의 실행 시간을 **20분 내로 단축**하고 **RMSE를 최소화**하여 경진대회 제출 점수를 극대화합니다.

제시된 전략에 따라, 아래의 구체적인 실행 계획을 수립하고 진행 상황을 체크합니다.

---

### **Ⅰ. 모델링 전략 수립 (Modeling Strategy)**

- [x] **단일 모델(LightGBM) 집중 전략 채택**
  - **이유:** 4개 모델(LGBM, CatBoost, RF 등)을 모두 사용하는 것은 20분 내 실행이 불가능. LightGBM은 속도와 성능이 가장 균형 잡혀 있으므로, 피처 엔지니어링 및 하이퍼파라미터 튜닝에 시간을 집중 투자하는 것이 효율적.
- [x] (선택) **듀얼 모델(LGBM + CatBoost) 블렌딩 앙상블 시도**
  - LightGBM 단일 모델의 성능이 한계에 도달했을 경우, 빠른 모델 2개를 조합하는 블렌딩(가중 평균) 전략을 고려합니다.

---

### **Ⅱ. 실행 시간 단축 (Execution Time Reduction)**

- [x] **하이퍼파라미터 튜닝(Optuna) 최적화**
    - [x] Optuna의 탐색 횟수 `n_trials`를 **15~20회**로 제한하여 튜닝 시간을 대폭 단축합니다.
    - [x] 튜닝 시, TimeSeriesSplit 전체가 아닌 **고정된 단일 검증 세트(Fixed Validation Set)**를 사용하여 최적화 속도를 높입니다.
- [x] **교차 검증(CV) 폴드 수 축소**
    - [x] `TimeSeriesSplit`의 `n_splits` 파라미터를 기존 5에서 **3 또는 4**로 줄여 전체 학습 및 검증 시간을 단축합니다.
- [x] **적극적인 피처 선택(Feature Selection) 수행**
    - [x] 1차 학습을 통해 LightGBM의 `feature_importance_`를 계산합니다.
    - [x] 중요도 기준 **상위 50~100개**의 핵심 피처만 선택하여 최종 모델링에 사용합니다.
- [x] **개발 단계에서 데이터 샘플링 활용**
    - [x] 코드 개발 및 테스트 단계에서는 `train.sample(frac=0.5)` 와 같이 데이터를 샘플링하여 빠른 피드백 루프를 구축합니다.
    - [x] 최종 제출 파일을 생성할 때만 전체 데이터를 사용하도록 파이프라인을 구성합니다.

---

### **Ⅲ. RMSE 개선을 위한 피처 엔지니어링 (Feature Engineering for RMSE Improvement)**

- [x] **타겟 변수 로그 변환**
    - [x] `target` 변수에 `np.log1p`를 적용하여 정규성을 확보하고, 예측 후 `np.expm1`으로 복원하는 로직을 유지합니다.
- [x] **시차 피처(Lag Features) 생성 (핵심)**
    - [x] '법정동' 등 주요 카테고리를 기준으로 **1개월, 3개월, 6개월 전**의 평균 가격 피처를 생성합니다. 과거의 가격 정보는 현재 가격 예측에 매우 중요한 정보입니다.
- [x] **이동 평균 피처(Rolling Window Features) 생성 (핵심)**
    - [x] '법정동' 등 주요 카테고리를 기준으로 최근 **3개월, 6개월간 가격의 이동 평균, 이동 표준편차, 이동 최소/최대값**을 피처로 추가하여 가격의 추세와 변동성을 모델에 학습시킵니다.
- [x] **기존 성공 피처 재활용 및 검토**
    - [x] K-Means 군집화 피처
    - [x] 주기성(Cyclical) 피처 (월, 분기 등 시간 관련 정보)
    - [x] 상호작용 피처 (예: 총면적 * 건물 연식, 층수 / 최고 층수 등)

---

### **Ⅳ. 전체 파이프라인 구현 계획 (Implementation Pipeline)**

1.  **[ ] 1. 기본 설정 및 데이터 로딩**
    - [x] 라이브러리 임포트 및 데이터셋 로드
    - [x] 타겟 변수에 `np.log1p` 적용
2.  **[ ] 2. 핵심 피처 엔지니어링**
    - [x] `Ⅲ`번 항목의 피처 엔지니어링(Lag, Rolling Window 등)을 모두 수행합니다.
3.  **[ ] 3. 피처 선택 및 데이터셋 준비**
    - [x] 생성된 전체 피처를 바탕으로 LightGBM을 빠르게 학습시켜 피처 중요도를 계산합니다.
    - [x] 중요도 상위 80개(예시) 피처를 선택하여 최종 학습 데이터셋(`X_final`, `y_final`)을 구성합니다.
4.  **[ ] 4. 모델 학습 및 검증 설정**
    - [x] 교차 검증 전략으로 `TimeSeriesSplit(n_splits=3)`을 설정합니다.
5.  **[ ] 5. 하이퍼파라미터 튜닝**
    - [x] Optuna와 고정 검증 세트를 사용하여 LightGBM 하이퍼파라미터를 `n_trials=20`으로 빠르게 튜닝합니다.
6.  **[ ] 6. 최종 모델 학습 및 예측**
    - [x] 튜닝된 최적 파라미터와 `TimeSeriesSplit(n_splits=3)`을 사용하여 최종 모델을 학습합니다.
    - [x] 테스트 데이터에 대한 예측을 수행합니다.
    - [x] 예측 결과에 `np.expm1`을 적용하여 스케일을 복원합니다.
7.  **[ ] 7. 제출 및 시간 측정**
    - [x] 최종 제출 파일을 생성합니다.
    - [x] 전체 파이프라인의 실행 시간이 20분 이내인지 확인하고 기록합니다.

**Comment by iejob** [comment]
- 피쳐 엔지니어링 후 시각화
<img width="1005" height="633" alt="Image" src="https://github.com/user-attachments/assets/53e08ceb-f113-4f9c-a012-e724d1698294" />
<img width="928" height="829" alt="Image" src="https://github.com/user-attachments/assets/94f28658-eed8-406f-9c78-4c9887635ec8" />
<img width="1784" height="483" alt="Image" src="https://github.com/user-attachments/assets/1146e548-3fff-486b-96a4-e9a5e79a2e69" />


---

- 상위 피쳐 25개 선택 후 시각화
<img width="1486" height="384" alt="Image" src="https://github.com/user-attachments/assets/e4a0f4c5-ce16-4b13-b79a-6785b6becf77" />
<img width="2698" height="2789" alt="Image" src="https://github.com/user-attachments/assets/d63872e4-04cc-4ddc-97b6-f8add738c0ce" />

---

- 모델 학습 완료 시각화
<img width="855" height="867" alt="Image" src="https://github.com/user-attachments/assets/206cf9c4-44db-4572-9005-7df8a7c04ba1" />
<img width="865" height="559" alt="Image" src="https://github.com/user-attachments/assets/b4af7539-8695-43bd-88e7-17b1dcb1d2b1" />
<img width="871" height="559" alt="Image" src="https://github.com/user-attachments/assets/913f74d8-7b48-4af2-b65f-0441b91b1b10" />
<img width="864" height="558" alt="Image" src="https://github.com/user-attachments/assets/0b0651b9-98a1-429e-9421-28d0177bc521" />
<img width="1181" height="841" alt="Image" src="https://github.com/user-attachments/assets/2c726cc6-4f95-4b19-8021-328508f84519" />
<img width="757" height="953" alt="Image" src="https://github.com/user-attachments/assets/79ce6c49-bf3c-4edf-b44c-840b62ebd581" />
<img width="695" height="477" alt="Image" src="https://github.com/user-attachments/assets/7703c59a-1057-4b23-b6b6-18d1aafff19e" />
<img width="718" height="484" alt="Image" src="https://github.com/user-attachments/assets/132c5370-594f-4f84-9d2d-76ba6cb9b603" />
<img width="695" height="477" alt="Image" src="https://github.com/user-attachments/assets/92c6ad45-2138-4f4c-9207-14ff4e7940ff" />
<img width="864" height="557" alt="Image" src="https://github.com/user-attachments/assets/33f1cb0f-1c37-4f0c-8f6e-80db13f65eec" />
